horovod/common/operations.cc

```diff
@@ -912,146 +347,58 @@ void CoordinateCacheAndState(CacheCoordinator& cache_coordinator,
 //      in the same order, so we cannot dispatch one thread per tensor,
 //      otherwise we may end up dispatching many blocked threads and never make
 //      progress if we have a thread pool limit.
-bool RunLoopOnce(HorovodGlobalState& state, MPIContext& ctx,
-                 bool is_coordinator);
-
-void BackgroundThreadLoop(HorovodGlobalState& state, MPIContext& ctx) {
-  // Initialize MPI if it was not initialized. This must happen on the
-  // background thread, since not all MPI implementations support being called
-  // from multiple threads.
-  //
-  // In some cases MPI library has multi-threading support, but it slows down
-  // certain components, e.g. OpenIB BTL in OpenMPI gets disabled if
-  // MPI_THREAD_MULTIPLE is requested.
-  //
-  // By default, we will ask for multiple threads, so other libraries like
-  // mpi4py can be used together with Horovod if multi-threaded MPI is
-  // installed.
-  auto mpi_threads_disable = std::getenv(HOROVOD_MPI_THREADS_DISABLE);
-  int required = MPI_THREAD_MULTIPLE;
-  if (mpi_threads_disable != nullptr &&
-      std::strtol(mpi_threads_disable, nullptr, 10) > 0) {
-    required = MPI_THREAD_SINGLE;
-  }
-  int provided;
-  int is_mpi_initialized = 0;
-  MPI_Initialized(&is_mpi_initialized);
-  if (is_mpi_initialized) {
-    MPI_Query_thread(&provided);
-    if (provided < MPI_THREAD_MULTIPLE) {
-      LOG(WARNING)
-          << "MPI has already been initialized without "
-             "multi-threading support (MPI_THREAD_MULTIPLE). This will "
-             "likely cause a segmentation fault.";
-    }
-  } else {
+bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator);
+
+void BackgroundThreadLoop(HorovodGlobalState& state) {
+
+  state.cpu_operation = ParseCPUOpsFromEnv();
+
 #if HAVE_DDL
-    // DDL comes with IBM Spectrum MPI
-    // and needs to initialize MPI with the proper license.
-    DDLAllreduce::DDLInit(&ddl_context, &cuda_context);
+  // If DDL is enabled, let DDL ops manage MPI environment.
```

Comment:
I guess since this `mpi_ctx_manager` is only used in this function, we don't need to make it a shared pointer.  We can just create it on the stack.

Reply:
That's right.